# c-di-GMP_host-association

This is the github repository for a study describing the key role of bacterial c-di-GMP in the evolution of host-association. 

## Abstract
Most microbes evolve faster than their hosts and should therefore drive evolution of host-microbe interactions. However, relatively little is known about the characteristics that define the adaptive path of microbes to host-association. In this study we have identified microbial traits that mediate adaptation to hosts by experimentally evolving the bacterium *Pseudomonas lurida* with the nematode *Caenorhabditis elegans*. We repeatedly observed the evolution of beneficial host-specialist bacteria with improved persistence in the nematode, achieved by mutations that uniformly upregulate the universal second messenger c-di-GMP. We subsequently upregulated c-di-GMP in different *Pseudomonas* species, consistently causing increased host-association. Comparison of Pseudomonad genomes from various environments revealed that c-di-GMP underlies adaptation to a variety of hosts, from plants to humans, suggesting that it is fundamental for establishing host-association.

*From Obeng et al. (2023) bioRxiv, doi: 10.1101/2023.03.20.533436*

## Description of files
In this repository, we provide  data and corresponding analysis scripts used to prepare the manuscript figures. Sub-directories are divided by approaches, i.e. experimental data and analysis ("ExperimentalAnalyses"), genomic analysis ("Genomics") and mathematical modeling ("MathematicalModeling"). We further provide plasmid maps used in our study in the respective directory ("Plasmids").

## Authors
Nancy Obeng, Anna Czerwinski, Daniel Schütz, Jan Michels, Jan Leipert, Florence Bansept, María J. García García, Thekla Schultheiß, Melinda Kemlein, Janina Fuß, Andreas Tholey, Arne Traulsen, Holger Sondermann, Hinrich Schulenburg

## Contributions
N.O.: conceptualization, methodology, investigation, analysis, writing, supervision. A.C.: conceptualization, methodology, investigation, analysis, writing. D.S.: analysis, writing. J.M.: investigation, writing. J.L.: investigation, writing. F.B.: conceptualization, modelling, analysis, writing. M.J.G.G.: methodology, writing. T.S.: investigation, writing. M.K.: investigation, writing. J.F.: investigation, writing. A.Th.: writing, supervision. A.Tr.: conceptualization, writing, supervision. H.So.: methodology, writing.  H.Schu.: conceptualization, writing, supervision.

## Acknowledgements
We thank E. Stukenbrock for access to the LSM 880; K. Guillemin, P. Rainey, H. Schweizer, F. Yildiz and G. O’Toole for providing bacterial strains or plasmids; D. Rogers, J. Summers for guidance in allelic exchange; B. Pees for illustration support; S. Joel, J. Hofmann, J. Löwenstrom, J. Lorenzen, H. Griem-Krey, L. Bluhm, and L. Rheindorf for lab support, the Kiel BiMo/LMB for access to their core facilities, the Schulenburg lab for project feedback, and B. Bohannan, R. Knight, and P. Engel for advice on the manuscript. Funding was provided by the Deutsche Forschungsgemeinschaft (DFG, German Research Foundation), Project-ID 261376515 – SFB 1182, Projects A4 and Z3 (NO, AC, FB, JL, A Tho, A Tr, H Schu), the DFG Research Infrastructure NGS_CC project 407495230 (JF) as part of the Next Generation Sequencing Competence Network project 423957469, the International Max-Planck Research School for Evolutionary Biology (NO, AC), the Max-Planck Society (Fellowship to H Sch), and NIH project R01AI168017 (MJGG, H So).
